Kitodo-FileManagement/src/main/java/org/kitodo/filemanagement/locking/LockManagement.java

```diff
@@ -0,0 +1,494 @@
+ */
+
+package org.kitodo.filemanagement.locking;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.UncheckedIOException;
+import java.net.ProtocolException;
+import java.net.URI;
+import java.nio.file.AccessDeniedException;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.kitodo.api.filemanagement.LockingMode;
+import org.kitodo.api.filemanagement.LockingResult;
+
+/**
+ * This is the main class of lock management. In regular operation, there must
+ * always be just one lock management for comprehensible reasons, for which the
+ * class implements the singleton pattern. The lock management has been divided
+ * into three classes for the sake of clarity, the actual lock management is in
+ * this class and manages the locks as such. The stream management manages the
+ * input and output streams flowing through the locks, and the immutable read
+ * file management manages the temporary files that must be generated for the
+ * immutable read lock and later deleted.
+ */
+public class LockManagement {
+    /**
+     * The singleton instance of the lock management. This class should be used
+     * as singleton, meaning that there is only one lock management all over the
+     * class loader. Everything else would not make sense or would even be
+     * dangerous. Nevertheless, the methods were not simply implemented
+     * {@code static}ally, because in the tests for this class, lock management
+     * must be re-initialized for each test. But for the usage, only the
+     * singleton should be used and no more instances of the class should exist.
+     */
+    private static volatile LockManagement instance;
+
+    private static final Logger logger = LogManager.getLogger(LockManagement.class);
+
+    /**
+     * This message is used in exceptions when a wrong object has been
+     * submitted.
+     */
+    private static final String UNSUITABLE_LOCKING_RESULT = "The locking result is unsuitable for generating the stream: ";
+
+    /**
+     * Returns the singleton instance of the lock management. This class should
+     * be used as singleton, meaning that there is only one lock management all
+     * over the class loader. For the usage, only the singleton should be used
+     * and no more instances of the class should exist. Everything else would
+     * not make sense or would even be dangerous.
+     *
+     * @return the singleton instance of this class
+     *
+     * @see "https://en.wikipedia.org/wiki/Double-checked_locking"
+     */
+    public static LockManagement singleton() {
```

Comment:
Can you please use the standard "getInstance" method for the singleton?
I think a private constructore is missing here too

Reply:
I renamed the getter. A private constructor is not possible because otherwise the tests will not be able to get their own instance of the class. In operation, the class should function as a singleton, but in the tests each test must be able to get a newly initialized lock management.

Reply:
But that's not an acceptable reason. If each tests needs a newly initialized lock management that you have to write your tests or class accordingly. That is definitely possible.

You could implement a `clear`-method so you're able to wipe the instance if necessary. Then you can set up a new one for each test. Singleton doesn't mean that it has to stay around forever but that there can't be multiple ones at the same time.

Reply:
If the lock management has a clear method that is not better, you can also use it wrong. If possible, I would also like to avoid having methods in the productive code that are only used by tests. But if you care so much I can change it.

Reply:
> I would also like to avoid having methods in the productive code that are only used by tests

The only reason for violating the private constructor idea is that your tests should work. You're doing what you don't want to do as well.

I'd prefer it if the tests were independent and wouldn't need a new instance every time. But as this is not happening with the Singleton pattern a `clear`-method should be preferred. That's a clean interface that's not dictated by the tests.

> you can also use it wrong

But that's what exactly what code review is for. So that you don't use it wrong.😛